httpcore/_backends/sync.py

```diff
@@ -17,6 +19,117 @@
 from .base import SOCKET_OPTION, NetworkBackend, NetworkStream
 
 
+class SyncTLSStream(NetworkStream):
+    """
+    Because the standard `SSLContext.wrap_socket` method does
+    not work for `SSLSocket` objects, we need this class
+    to implement TLS stream using an underlying `SSLObject`
+    instance in order to support TLS on top of TLS.
+    """
+
+    # Defined in RFC 8449
+    TLS_RECORD_SIZE = 16384
+
+    def __init__(
+        self,
+        sock: socket.socket,
+        ssl_context: ssl.SSLContext,
+        server_hostname: typing.Optional[str] = None,
+        timeout: typing.Optional[float] = None,
+    ):
+        self._sock = sock
+        self._incoming = ssl.MemoryBIO()
+        self._outgoing = ssl.MemoryBIO()
+
+        self.ssl_obj = ssl_context.wrap_bio(
+            incoming=self._incoming,
+            outgoing=self._outgoing,
+            server_hostname=server_hostname,
+        )
+
+        self._perform_io(self.ssl_obj.do_handshake, timeout)
+
+    def _perform_io(
+        self,
+        func: typing.Callable[..., typing.Any],
+        timeout: typing.Optional[float],
+    ) -> typing.Any:
+        ret = None
+        timeout = timeout or None  # Replaces `0` with `None`
```

Comment:
?

Reply:
We need to distinguish between cases where we got 0 after decreasing our timeout and cases where we don't want to handle timeout at all.

Reply:
if given paramter's value was `0.0`, it replaces it with `None` to avoid `socket.timeout` at [here](https://github.com/encode/httpcore/pull/732/files#diff-d3a40cf09f980d710bfc8d32385d11086ebd179cec344a9ac5c095251f7b6224R67)